lib/scanner/walk.go

```diff
@@ -242,40 +274,50 @@ func (w *walker) walkAndHashFiles(ctx context.Context, fchan, dchan chan protoco
 			return skip
 		}
 
-		if w.Matcher.Match(path).IsIgnored() {
-			l.Debugln("ignored (patterns):", path)
-			return skip
-		}
-
 		if !utf8.ValidString(path) {
 			l.Warnf("File name %q is not in UTF8 encoding; skipping.", path)
 			return skip
 		}
 
-		path, shouldSkip := w.normalizePath(path, info)
-		if shouldSkip {
-			return skip
-		}
-
-		switch {
-		case info.IsSymlink():
-			if err := w.walkSymlink(ctx, path, dchan); err != nil {
-				return err
+		if w.Matcher.Match(path).IsIgnored() {
+			l.Debugln("ignored (patterns):", path)
+			if w.Matcher.SkipIgnoredDirs() {
+				return skip
 			}
-			if info.IsDir() {
-				// under no circumstances shall we descend into a symlink
-				return fs.SkipDir
+			// If the parent wasn't ignored already, set this path as the "highest" ignored parent
+			if ignoredParent == "" || !strings.HasPrefix(path, ignoredParent+string(fs.PathSeparator)) {
```

Comment:
Shouldn't we only set this if path is a directory?

Reply:
No, the "upmost" parent is kept, because at this point we don't yet know whether there is an unignored child. And if there is, we need to handle all parents.

Reply:
You could do it the other way around, which would make it less magical.

Keep building up ignoredParent up to the unignored item, then once you do hit an unignored item, walk ignoredParent backwards up to the point where ignoredParent is not actually ignored.

Also the directory check thing still prevails.

Reply:
This doesn't work, we need to keep track of the upmost/shortest ignored directory. Consider the patterns
```
!not
/foo
```
and files `foo/aaa/ignored` and `foo/not`. `foo` is ignored and becomes `ignoredParent`, then `foo/aaa` becomes `ignoredParent`. `foo/aaa/ignored` is just ignored, nothing else happens. Then `foo/not` is not ignored, but `ignoredParents` is `foo/aaa` so `foo` is not handled.